File: codes/perl/Array.find.pl
use strict;
use warnings;
use JSON;
use Scalar::Util qw(looks_like_number);

# There's no JavaScript-like Array.find() in Perl.
# But, we can create our own function to mimic it in Perl.

sub pretty_array_of_primitives {
    my $number_of_parameters = @_;
    my $result = "[";
    for (my $array_item_index = 0; $array_item_index < $number_of_parameters; $array_item_index += 1) {
        my $array_item = $_[$array_item_index];
        my $is_string = defined($array_item) && $array_item =~ /[0-9a-zA-Z`~!@#%&_=;':", \(\)\[\]\{\}\|\*\+\?\^\$\/\\\<\>\.\-]/;
        my $is_number = looks_like_number($array_item);
        last if (!$is_string && !$is_number);
        $result = $result . "\"" . $array_item . "\"" if ($is_string);
        $result = $result . $array_item if ($is_number);
        $result = $result . ", " if (($array_item_index + 1) != $number_of_parameters);
    }
    $result = $result . "]";
    return $result;
}

sub array_find_v1 {
    # JavaScript-like Array.find() function
    my ($callback_function, @an_array) = @_;
    my $data_found;
    for my $array_item_index (0..$#an_array) {
        my $array_item = $an_array[$array_item_index];
        my $is_condition_match = $callback_function->($array_item, $array_item_index, \@an_array);
        if ($is_condition_match) {
            $data_found = $array_item;
            last;
        }
    }
    return $data_found;
}

sub array_find_v2 {
    # JavaScript-like Array.find() function
    my ($callback_function, @an_array) = @_;
    my $data_found;
    for my $array_item_index (0..$#an_array) {
        my $array_item = $an_array[$array_item_index];
        if ($callback_function->($array_item, $array_item_index, \@an_array)) {
            $data_found = $array_item;
            last;
        }
    }
    return $data_found;
}

sub array_find_v3 {
    # JavaScript-like Array.find() function
    my ($callback_function, @an_array) = @_;
    my $data_found;
    for my $array_item_index (0..$#an_array) {
        my $array_item = $an_array[$array_item_index];
        my $is_condition_match = $callback_function->($array_item, $array_item_index, \@an_array);
        if ($is_condition_match) {
            return $array_item;
        }
    }
    return $data_found;
}

sub array_find_v4 {
    # JavaScript-like Array.find() function
    my ($callback_function, @an_array) = @_;
    my $data_found;
    for my $array_item_index (0..$#an_array) {
        my $array_item = $an_array[$array_item_index];
        my $is_condition_match = $callback_function->($array_item, $array_item_index, \@an_array);
        return $array_item if ($is_condition_match);
    }
    return $data_found;
}

sub array_find_v5 {
    # JavaScript-like Array.find() function
    my ($callback_function, @an_array) = @_;
    my $data_found;
    for my $array_item_index (0..$#an_array) {
        my $array_item = $an_array[$array_item_index];
        if ($callback_function->($array_item, $array_item_index, \@an_array)) {
            return $array_item;
        }
    }
    return $data_found;
}

sub array_find_v6 {
    # JavaScript-like Array.find() function
    my ($callback_function, @an_array) = @_;
    my $data_found;
    for my $array_item_index (0..$#an_array) {
        my $array_item = $an_array[$array_item_index];
        return $array_item if ($callback_function->($array_item, $array_item_index, \@an_array));
    }
    return $data_found;
}

sub array_find_v7 {
    # JavaScript-like Array.find() function
    my ($callback_function, @an_array) = @_;
    for my $array_item_index (0..$#an_array) {
        my $array_item = $an_array[$array_item_index];
        my $is_condition_match = $callback_function->($array_item, $array_item_index, \@an_array);
        if ($is_condition_match) {
            return $array_item;
        }
    }
    return undef;
}

sub array_find_v8 {
    # JavaScript-like Array.find() function
    my ($callback_function, @an_array) = @_;
    for my $array_item_index (0..$#an_array) {
        my $array_item = $an_array[$array_item_index];
        my $is_condition_match = $callback_function->($array_item, $array_item_index, \@an_array);
        return $array_item if ($is_condition_match);
    }
    return undef;
}

sub array_find_v9 {
    # JavaScript-like Array.find() function
    my ($callback_function, @an_array) = @_;
    for my $array_item_index (0..$#an_array) {
        my $array_item = $an_array[$array_item_index];
        if ($callback_function->($array_item, $array_item_index, \@an_array)) {
            return $array_item;
        }
    }
    return undef;
}

sub array_find_v10 {
    # JavaScript-like Array.find() function
    my ($callback_function, @an_array) = @_;
    for my $array_item_index (0..$#an_array) {
        my $array_item = $an_array[$array_item_index];
        return $array_item if ($callback_function->($array_item, $array_item_index, \@an_array));
    }
    return undef;
}

print("\n# JavaScript-like Array.find() in Perl Array\n");

my @numbers = (12, 34, 27, 23, 65, 93, 36, 87, 4, 254);
print("numbers: ", pretty_array_of_primitives(@numbers), "\n");

my $even_number_found;
my $odd_number_found;

print("# using JavaScript-like Array.find() function \"array_find_v1\"\n");

$even_number_found = array_find_v1(sub { my ($number) = @_; return (($number % 2) == 0); }, @numbers);
print("even number found: ", $even_number_found, "\n");
# even number found: 12

$odd_number_found = array_find_v1(sub { my ($number) = @_; return (($number % 2) != 0); }, @numbers);
print("odd number found: ", $odd_number_found, "\n");
# odd number found: 27

print("# using JavaScript-like Array.find() function \"array_find_v2\"\n");

$even_number_found = array_find_v2(sub { my ($number) = @_; return (($number % 2) == 0); }, @numbers);
print("even number found: ", $even_number_found, "\n");
# even number found: 12

$odd_number_found = array_find_v2(sub { my ($number) = @_; return (($number % 2) != 0); }, @numbers);
print("odd number found: ", $odd_number_found, "\n");
# odd number found: 27

print("# using JavaScript-like Array.find() function \"array_find_v3\"\n");

$even_number_found = array_find_v3(sub { my ($number) = @_; return (($number % 2) == 0); }, @numbers);
print("even number found: ", $even_number_found, "\n");
# even number found: 12

$odd_number_found = array_find_v3(sub { my ($number) = @_; return (($number % 2) != 0); }, @numbers);
print("odd number found: ", $odd_number_found, "\n");
# odd number found: 27

print("# using JavaScript-like Array.find() function \"array_find_v4\"\n");

$even_number_found = array_find_v4(sub { my ($number) = @_; return (($number % 2) == 0); }, @numbers);
print("even number found: ", $even_number_found, "\n");
# even number found: 12

$odd_number_found = array_find_v4(sub { my ($number) = @_; return (($number % 2) != 0); }, @numbers);
print("odd number found: ", $odd_number_found, "\n");
# odd number found: 27

print("# using JavaScript-like Array.find() function \"array_find_v5\"\n");

$even_number_found = array_find_v5(sub { my ($number) = @_; return (($number % 2) == 0); }, @numbers);
print("even number found: ", $even_number_found, "\n");
# even number found: 12

$odd_number_found = array_find_v5(sub { my ($number) = @_; return (($number % 2) != 0); }, @numbers);
print("odd number found: ", $odd_number_found, "\n");
# odd number found: 27

print("# using JavaScript-like Array.find() function \"array_find_v6\"\n");

$even_number_found = array_find_v6(sub { my ($number) = @_; return (($number % 2) == 0); }, @numbers);
print("even number found: ", $even_number_found, "\n");
# even number found: 12

$odd_number_found = array_find_v6(sub { my ($number) = @_; return (($number % 2) != 0); }, @numbers);
print("odd number found: ", $odd_number_found, "\n");
# odd number found: 27

print("# using JavaScript-like Array.find() function \"array_find_v7\"\n");

$even_number_found = array_find_v7(sub { my ($number) = @_; return (($number % 2) == 0); }, @numbers);
print("even number found: ", $even_number_found, "\n");
# even number found: 12

$odd_number_found = array_find_v7(sub { my ($number) = @_; return (($number % 2) != 0); }, @numbers);
print("odd number found: ", $odd_number_found, "\n");
# odd number found: 27

print("# using JavaScript-like Array.find() function \"array_find_v8\"\n");

$even_number_found = array_find_v8(sub { my ($number) = @_; return (($number % 2) == 0); }, @numbers);
print("even number found: ", $even_number_found, "\n");
# even number found: 12

$odd_number_found = array_find_v8(sub { my ($number) = @_; return (($number % 2) != 0); }, @numbers);
print("odd number found: ", $odd_number_found, "\n");
# odd number found: 27

print("# using JavaScript-like Array.find() function \"array_find_v9\"\n");

$even_number_found = array_find_v9(sub { my ($number) = @_; return (($number % 2) == 0); }, @numbers);
print("even number found: ", $even_number_found, "\n");
# even number found: 12

$odd_number_found = array_find_v9(sub { my ($number) = @_; return (($number % 2) != 0); }, @numbers);
print("odd number found: ", $odd_number_found, "\n");
# odd number found: 27

print("# using JavaScript-like Array.find() function \"array_find_v10\"\n");

$even_number_found = array_find_v10(sub { my ($number) = @_; return (($number % 2) == 0); }, @numbers);
print("even number found: ", $even_number_found, "\n");
# even number found: 12

$odd_number_found = array_find_v10(sub { my ($number) = @_; return (($number % 2) != 0); }, @numbers);
print("odd number found: ", $odd_number_found, "\n");
# odd number found: 27

print("\n# JavaScript-like Array.find() in Perl Array of Hashes\n");

my @products = (
    {
        "code" => "pasta",
        "price" => 321
    },
    {
        "code" => "bubble_gum",
        "price" => 233
    },
    {
        "code" => "potato_chips",
        "price" => 5
    },
    {
        "code" => "towel",
        "price" => 499
    }
);

print("products: ", JSON->new->allow_nonref->pretty->encode(\@products));

my $product_found;

print("# using JavaScript-like Array.find() function \"array_find_v1\"\n");

$product_found = array_find_v1(sub { my ($product) = @_; return ($product->{"code"} eq "bubble_gum"); }, @products);
print("product found: ", defined($product_found) ? JSON->new->allow_nonref->pretty->encode(\%$product_found) : "undefined\n");
# product found: {
#     "code": "bubble_gum",
#     "price": 233
# }

print("# using JavaScript-like Array.find() function \"array_find_v2\"\n");

$product_found = array_find_v2(sub { my ($product) = @_; return ($product->{"code"} eq "bubble_gum"); }, @products);
print("product found: ", defined($product_found) ? JSON->new->allow_nonref->pretty->encode(\%$product_found) : "undefined\n");
# product found: {
#     "code": "bubble_gum",
#     "price": 233
# }

print("# using JavaScript-like Array.find() function \"array_find_v3\"\n");

$product_found = array_find_v3(sub { my ($product) = @_; return ($product->{"code"} eq "bubble_gum"); }, @products);
print("product found: ", defined($product_found) ? JSON->new->allow_nonref->pretty->encode(\%$product_found) : "undefined\n");
# product found: {
#     "code": "bubble_gum",
#     "price": 233
# }

print("# using JavaScript-like Array.find() function \"array_find_v4\"\n");

$product_found = array_find_v4(sub { my ($product) = @_; return ($product->{"code"} eq "bubble_gum"); }, @products);
print("product found: ", defined($product_found) ? JSON->new->allow_nonref->pretty->encode(\%$product_found) : "undefined\n");
# product found: {
#     "code": "bubble_gum",
#     "price": 233
# }

print("# using JavaScript-like Array.find() function \"array_find_v5\"\n");

$product_found = array_find_v5(sub { my ($product) = @_; return ($product->{"code"} eq "bubble_gum"); }, @products);
print("product found: ", defined($product_found) ? JSON->new->allow_nonref->pretty->encode(\%$product_found) : "undefined\n");
# product found: {
#     "code": "bubble_gum",
#     "price": 233
# }

print("# using JavaScript-like Array.find() function \"array_find_v6\"\n");

$product_found = array_find_v6(sub { my ($product) = @_; return ($product->{"code"} eq "bubble_gum"); }, @products);
print("product found: ", defined($product_found) ? JSON->new->allow_nonref->pretty->encode(\%$product_found) : "undefined\n");
# product found: {
#     "code": "bubble_gum",
#     "price": 233
# }

print("# using JavaScript-like Array.find() function \"array_find_v7\"\n");

$product_found = array_find_v7(sub { my ($product) = @_; return ($product->{"code"} eq "bubble_gum"); }, @products);
print("product found: ", defined($product_found) ? JSON->new->allow_nonref->pretty->encode(\%$product_found) : "undefined\n");
# product found: {
#     "code": "bubble_gum",
#     "price": 233
# }

print("# using JavaScript-like Array.find() function \"array_find_v8\"\n");

$product_found = array_find_v8(sub { my ($product) = @_; return ($product->{"code"} eq "bubble_gum"); }, @products);
print("product found: ", defined($product_found) ? JSON->new->allow_nonref->pretty->encode(\%$product_found) : "undefined\n");
# product found: {
#     "code": "bubble_gum",
#     "price": 233
# }

print("# using JavaScript-like Array.find() function \"array_find_v9\"\n");

$product_found = array_find_v9(sub { my ($product) = @_; return ($product->{"code"} eq "bubble_gum"); }, @products);
print("product found: ", defined($product_found) ? JSON->new->allow_nonref->pretty->encode(\%$product_found) : "undefined\n");
# product found: {
#     "code": "bubble_gum",
#     "price": 233
# }

print("# using JavaScript-like Array.find() function \"array_find_v10\"\n");

$product_found = array_find_v10(sub { my ($product) = @_; return ($product->{"code"} eq "bubble_gum"); }, @products);
print("product found: ", defined($product_found) ? JSON->new->allow_nonref->pretty->encode(\%$product_found) : "undefined\n");
# product found: {
#     "code": "bubble_gum",
#     "price": 233
# }

File: codes/ruby/array_find.rb
require "json"

array_find_v1 = ->(callback_function, an_array) do
    # JavaScript-like Array.find() function
    data_found = nil
    an_array.each_with_index do |array_item, array_item_index|
        is_condition_match = callback_function.call(array_item, array_item_index, an_array)
        if (is_condition_match == true)
            data_found = array_item
            break
        end
    end
    return data_found
end

array_find_v2 = ->(callback_function, an_array) do
    # JavaScript-like Array.find() function
    data_found = nil
    an_array.each_with_index do |array_item, array_item_index|
        is_condition_match = callback_function.call(array_item, array_item_index, an_array)
        unless (is_condition_match == false)
            data_found = array_item
            break
        end
    end
    return data_found
end

array_find_v3 = ->(callback_function, an_array) do
    # JavaScript-like Array.find() function
    data_found = nil
    an_array.each_with_index do |array_item, array_item_index|
        if (callback_function.call(array_item, array_item_index, an_array) == true)
            data_found = array_item
            break
        end
    end
    return data_found
end

array_find_v4 = ->(callback_function, an_array) do
    # JavaScript-like Array.find() function
    data_found = nil
    an_array.each_with_index do |array_item, array_item_index|
        unless (callback_function.call(array_item, array_item_index, an_array) == false)
            data_found = array_item
            break
        end
    end
    return data_found
end

array_find_v5 = ->(callback_function, an_array) do
    # JavaScript-like Array.find() function
    an_array.each_with_index do |array_item, array_item_index|
        is_condition_match = callback_function.call(array_item, array_item_index, an_array)
        return array_item if (is_condition_match == true)
    end
    return nil
end

array_find_v6 = ->(callback_function, an_array) do
    # JavaScript-like Array.find() function
    an_array.each_with_index do |array_item, array_item_index|
        is_condition_match = callback_function.call(array_item, array_item_index, an_array)
        return array_item unless (is_condition_match == false)
    end
    return nil
end

array_find_v7 = ->(callback_function, an_array) do
    # JavaScript-like Array.find() function
    an_array.each_with_index do |array_item, array_item_index|
        return array_item if (callback_function.call(array_item, array_item_index, an_array) == true)
    end
    return nil
end

array_find_v8 = ->(callback_function, an_array) do
    # JavaScript-like Array.find() function
    an_array.each_with_index do |array_item, array_item_index|
        return array_item unless (callback_function.call(array_item, array_item_index, an_array) == false)
    end
    return nil
end

print("\n", "# JavaScript-like Array.find() in Ruby Array")

numbers = [12, 34, 27, 23, 65, 93, 36, 87, 4, 254]
print("\n", "numbers: ", JSON.generate(numbers).gsub(",", ", "))

print("\n", "# using JavaScript-like Array.find() function \"array_find_v1\"")

even_number_found = array_find_v1.call(->(number, _, _) { ((number % 2) == 0) }, numbers)
print("\n", "even number found: ", even_number_found)
# even number found: 12

odd_numbers_found = array_find_v1.call(->(number, _, _) { ((number % 2) != 0) }, numbers)
print("\n", "odd number found: ", odd_numbers_found)
# odd number found: 27

print("\n", "# using JavaScript-like Array.find() function \"array_find_v2\"")

even_number_found = array_find_v2.call(->(number, _, _) { ((number % 2) == 0) }, numbers)
print("\n", "even number found: ", even_number_found)
# even number found: 12

odd_numbers_found = array_find_v2.call(->(number, _, _) { ((number % 2) != 0) }, numbers)
print("\n", "odd number found: ", odd_numbers_found)
# odd number found: 27

print("\n", "# using JavaScript-like Array.find() function \"array_find_v3\"")

even_number_found = array_find_v3.call(->(number, _, _) { ((number % 2) == 0) }, numbers)
print("\n", "even number found: ", even_number_found)
# even number found: 12

odd_numbers_found = array_find_v3.call(->(number, _, _) { ((number % 2) != 0) }, numbers)
print("\n", "odd number found: ", odd_numbers_found)
# odd number found: 27

print("\n", "# using JavaScript-like Array.find() function \"array_find_v4\"")

even_number_found = array_find_v4.call(->(number, _, _) { ((number % 2) == 0) }, numbers)
print("\n", "even number found: ", even_number_found)
# even number found: 12

odd_numbers_found = array_find_v4.call(->(number, _, _) { ((number % 2) != 0) }, numbers)
print("\n", "odd number found: ", odd_numbers_found)
# odd number found: 27

print("\n", "# using JavaScript-like Array.find() function \"array_find_v5\"")

even_number_found = array_find_v5.call(->(number, _, _) { ((number % 2) == 0) }, numbers)
print("\n", "even number found: ", even_number_found)
# even number found: 12

odd_numbers_found = array_find_v5.call(->(number, _, _) { ((number % 2) != 0) }, numbers)
print("\n", "odd number found: ", odd_numbers_found)
# odd number found: 27

print("\n", "# using JavaScript-like Array.find() function \"array_find_v6\"")

even_number_found = array_find_v6.call(->(number, _, _) { ((number % 2) == 0) }, numbers)
print("\n", "even number found: ", even_number_found)
# even number found: 12

odd_numbers_found = array_find_v6.call(->(number, _, _) { ((number % 2) != 0) }, numbers)
print("\n", "odd number found: ", odd_numbers_found)
# odd number found: 27

print("\n", "# using JavaScript-like Array.find() function \"array_find_v7\"")

even_number_found = array_find_v7.call(->(number, _, _) { ((number % 2) == 0) }, numbers)
print("\n", "even number found: ", even_number_found)
# even number found: 12

odd_numbers_found = array_find_v7.call(->(number, _, _) { ((number % 2) != 0) }, numbers)
print("\n", "odd number found: ", odd_numbers_found)
# odd number found: 27

print("\n", "# using JavaScript-like Array.find() function \"array_find_v8\"")

even_number_found = array_find_v8.call(->(number, _, _) { ((number % 2) == 0) }, numbers)
print("\n", "even number found: ", even_number_found)
# even number found: 12

odd_numbers_found = array_find_v8.call(->(number, _, _) { ((number % 2) != 0) }, numbers)
print("\n", "odd number found: ", odd_numbers_found)
# odd number found: 27

print("\n", "# using Ruby Array.find() built-in method \"Array.find\"")

even_number_found = numbers.find { |number| ((number % 2) == 0) }
print("\n", "even number found: ", even_number_found)
# even number found: 12

odd_numbers_found = numbers.find { |number| ((number % 2) != 0) }
print("\n", "odd number found: ", odd_numbers_found)
# odd number found: 27

print("\n", "# JavaScript-like Array.find() in Ruby Array of Hashes")

products = [
    {
        "code" => "pasta",
        "price" => 321
    },
    {
        "code" => "bubble_gum",
        "price" => 233
    },
    {
        "code" => "potato_chips",
        "price" => 5
    },
    {
        "code" => "towel",
        "price" => 499
    }
]
print("\n", "products: ", JSON.pretty_generate(products, { "indent": " " * 4 }))

product_to_find = "bubble_gum"
print("\n", "product to find: #{product_to_find}")

print("\n", "# using JavaScript-like Array.find() function \"array_find_v1\"")

product_found = array_find_v1.call(->(product, _, _) { (product["code"] == product_to_find) }, products)
print("\n", "product found: ", JSON.pretty_generate(product_found, { "indent": " " * 4 }))
# product found: {
#     "code": "bubble_gum",
#     "price": 233
# }

print("\n", "# using JavaScript-like Array.find() function \"array_find_v2\"")

product_found = array_find_v2.call(->(product, _, _) { (product["code"] == product_to_find) }, products)
print("\n", "product found: ", JSON.pretty_generate(product_found, { "indent": " " * 4 }))
# product found: {
#     "code": "bubble_gum",
#     "price": 233
# }

print("\n", "# using JavaScript-like Array.find() function \"array_find_v3\"")

product_found = array_find_v3.call(->(product, _, _) { (product["code"] == product_to_find) }, products)
print("\n", "product found: ", JSON.pretty_generate(product_found, { "indent": " " * 4 }))
# product found: {
#     "code": "bubble_gum",
#     "price": 233
# }

print("\n", "# using JavaScript-like Array.find() function \"array_find_v4\"")

product_found = array_find_v4.call(->(product, _, _) { (product["code"] == product_to_find) }, products)
print("\n", "product found: ", JSON.pretty_generate(product_found, { "indent": " " * 4 }))
# product found: {
#     "code": "bubble_gum",
#     "price": 233
# }

print("\n", "# using JavaScript-like Array.find() function \"array_find_v5\"")

product_found = array_find_v5.call(->(product, _, _) { (product["code"] == product_to_find) }, products)
print("\n", "product found: ", JSON.pretty_generate(product_found, { "indent": " " * 4 }))
# product found: {
#     "code": "bubble_gum",
#     "price": 233
# }

print("\n", "# using JavaScript-like Array.find() function \"array_find_v6\"")

product_found = array_find_v6.call(->(product, _, _) { (product["code"] == product_to_find) }, products)
print("\n", "product found: ", JSON.pretty_generate(product_found, { "indent": " " * 4 }))
# product found: {
#     "code": "bubble_gum",
#     "price": 233
# }

print("\n", "# using JavaScript-like Array.find() function \"array_find_v7\"")

product_found = array_find_v7.call(->(product, _, _) { (product["code"] == product_to_find) }, products)
print("\n", "product found: ", JSON.pretty_generate(product_found, { "indent": " " * 4 }))
# product found: {
#     "code": "bubble_gum",
#     "price": 233
# }

print("\n", "# using JavaScript-like Array.find() function \"array_find_v8\"")

product_found = array_find_v8.call(->(product, _, _) { (product["code"] == product_to_find) }, products)
print("\n", "product found: ", JSON.pretty_generate(product_found, { "indent": " " * 4 }))
# product found: {
#     "code": "bubble_gum",
#     "price": 233
# }

print("\n", "# using Ruby Array.find() built-in method \"Array.find\"")

product_found = products.find { |product| (product["code"] == product_to_find) }
print("\n", "product found: ", JSON.pretty_generate(product_found, { "indent": " " * 4 }))
# product found: {
#     "code": "bubble_gum",
#     "price": 233
# }
#
